### DISCUSSION TITLE
Más problemas SQL

### DISCUSSION THREAD

**Discussion by AstorArico72** [AstorArico72/Proyecto-Transversal-1#12] [General]
### Intenté ingresar datos a mis bases de datos SQL otra vez, pero hay problemas.
Parece que la consulta SQL es ejecutada 2 veces en mi método void.
Los datos son exitosamente ingresados (lo sé mediante otro método que muestra todo), pero después, la ejecución del método arroja una excepción SQL indicando que la clave primaria está duplicada, aunque en cada consulta SQL inserté una clave primaria diferente. Ésto es lo que sale al ejecutar el main:
`Excepción SQL encontrada: (conn=173) Duplicate entry '219' for key 'PRIMARY'` _Intenté con la clave primaria "219" (y otras más) en el método void nuevaMateria (Materia) para éste ejemplo._
Ésto pasa con el legajo del alumno también.
Aquí el código de mi método nuevaMateria (Materia). Decidí hacer múltiples catch como NetBeans sugería, y poner los mensajes en System.err.println.
```
public static void nuevaMateria (Materia unaMateria) { //Decidí hacerlo estático. Ver el hilo 10 para saber porqué.
        try {
            Class.forName ("org.mariadb.jdbc.Driver").getDeclaredConstructor ().newInstance ();
            conn = DriverManager.getConnection ("jdbc:mysql://localhost/Universidad", "root", "");
            String sqlQuery = "INSERT INTO Universidad.Materia (ID_Materia, Nombre, Año) VALUES (?, ?, ?)";
            ps = conn.prepareStatement (sqlQuery);
            ps.setInt (1, unaMateria.getID());
            ps.setString (2, unaMateria.getNombre());
            ps.setInt (3, unaMateria.getAño());
            ps.execute ();
            rs = ps.executeQuery();
            System.out.println ("Conexión exitosa.");
            System.out.println("Creada nueva materia #" + rs.getInt("ID_Materia") + ", año" + rs.getInt ("Año")); //Debería imprimir ésto...
        } catch (SQLException ex) {
            System.err.println ("Excepción SQL encontrada: " + ex.getMessage()); //...pero imprime ésto.
        } catch (ClassNotFoundException ex) {
            System.err.println ("Clase no encontrada: " + ex.getMessage());
        } catch (NoSuchMethodException ex) {
            System.err.println ("Método no encontrado: " + ex.getMessage());
        } catch (SecurityException ex) {
            System.err.println ("Excepción de seguridad encontrada: " + ex.getMessage());
        } catch (InstantiationException ex) {
            System.err.println ("Excepción de instanciación encontrada: " + ex.getMessage());
        } catch (IllegalAccessException ex) {
            System.err.println ("Acceso ilegal: " + ex.getMessage());
        } catch (IllegalArgumentException ex) {
            System.err.println ("Argumento ilegal: " + ex.getMessage());
        } catch (InvocationTargetException ex) {
            System.err.println ("Excepción de objetivo de invocamiento encontrada: " + ex.getMessage());
        }
    }
```
Es probable que hacer las claves primarias auto-incrementales arregle ésto.

**Comment by PabloBarrionuevo** [comment]
Astor en el proyecto hacele pull para que veas los cambios que vamos haciendo.
Las conexiones las dividimos así:
Existe una clase Conexión que va a hacer el registro en Class.forName(...) y va a establecer la conexión DriverManager.getConnection(...);
Luego están las clases data para trabajar con las tablas SQL de alumno, materia, ... etc.
Quedando algo así:
//Todos estos están el el paquete BD
BD.Conexion c = new BD.Conexion();
if(c.getConexion() != null){
                BD.MateriaData md = new BD.MateriaData(c); //le pasamos la conexion para que trabaje siempre con la misma
                int idNuevo = md.guardarMateria(nuevaMateria); //nuevaMateria ya existe en algun momento
                nuevaM.setIdMateria(idNuevo);
                //... etc
}
